ethcore/src/externalities.rs

```diff
@@ -131,20 +128,51 @@ impl<'a, T: 'a, V: 'a, B: 'a> Ext for Externalities<'a, T, V, B>
 		self.state.balance(address)
 	}
 
-	fn blockhash(&self, number: &U256) -> H256 {
-		// TODO: comment out what this function expects from env_info, since it will produce panics if the latter is inconsistent
-		match *number < U256::from(self.env_info.number) && number.low_u64() >= cmp::max(256, self.env_info.number) - 256 {
-			true => {
-				let index = self.env_info.number - number.low_u64() - 1;
-				assert!(index < self.env_info.last_hashes.len() as u64, format!("Inconsistent env_info, should contain at least {:?} last hashes", index+1));
-				let r = self.env_info.last_hashes[index as usize].clone();
-				trace!("ext: blockhash({}) -> {} self.env_info.number={}\n", number, r, self.env_info.number);
-				r
-			},
-			false => {
-				trace!("ext: blockhash({}) -> null self.env_info.number={}\n", number, self.env_info.number);
-				H256::zero()
-			},
+	fn blockhash(&mut self, number: &U256) -> H256 {
+		if self.env_info.number + 256 >= self.engine.params().eip210_transition {
+			let blockhash_contract_address = self.engine.params().eip210_contract_address;
+			let code_res = self.state.code(&blockhash_contract_address)
+				.and_then(|code| self.state.code_hash(&blockhash_contract_address).map(|hash| (code, hash)));
+
+			let (code, code_hash) = match code_res {
+				Ok((code, hash)) => (code, hash),
+				Err(_) => return H256::zero(),
+			};
+
```

Comment:
According to this comment https://github.com/ethereum/EIPs/pull/210#issuecomment-300357695 it was decided that `BLOCKHASH` instruction should return zero for blocks more-than-256 blocks old. So here should be a check on `number`.